### ISSUE TITLE
Rule Change: func-style: Allow arrow functions within functions

### ISSUE THREAD

**Issue by jeengbe** [eslint/eslint#18046]
### What rule do you want to change?

func-style

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
function doSomething() {
  const somethingElse = () => { ... };

  // ... stuff
}
```

### What does the rule currently do for this code?

Currently, the rule is all-in or all-out. Either disable the rule entirely, or change the inner code to:

```js
function doSomething() {
  function somethingElse() { ... }

  // ... stuff
}
```

With TypeScript, for instance, I've often seen patterns like

```ts
const doX: FunctionType = (a, b) => { ... };
```

Where `a`, `b` and the return type are inferred and checked automatically. This seldom happens at top-level functions.

### What will the rule do after it's changed?

I only care about functions at the top level. I would like the rule to enforce functions directly at the top (and immediate class methods) to be written as functions. Anything written within another function is subject to context and there is no one rule it's all kinda approach for those, imo. There are cases in which a function declaration within a function makes sense, other times, an arrow function is far more concise.

What also matters sometimes is that arrow functions don't rebind `this`, which sometimes makes them the only viable solution.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Thanks for the rule change proposal, but per our rules policies [stylistic rules are frozen](https://github.com/eslint/eslint#stylistic-rule-updates) so we are not adding new options to the `func-style` rule. You can always copy the existing rule into a custom rule and make it behave the way you want.
